language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs

```diff
@@ -207,6 +181,130 @@ newtype DynamicBuiltinNameMeanings = DynamicBuiltinNameMeanings
+
+Since type classes are globally coherent by design, we also have global coherence for dynamic built-in
+types for free. Any dynamic built-in type means the same thing regardless of the blockchain it's
+added to. It may prove to be restrictive, but it's a good property to start with, because less things
+can silently stab you in the back.
+
+An @KnownDynamicBuiltinType dyn@ instance provides
+
+1. a way to encode @dyn@ as a PLC type ('getTypeEncoding')
+2. a function that encodes values of type @dyn@ as PLC terms ('makeDynamicBuiltin')
+3. a function that decodes PLC terms back to Haskell values ('readDynamicBuiltin')
+
+The last two are ought to constitute an isomorphism (modulo 'Quote' and 'Maybe').
+-}
+
+{- Note [Converting PLC values to Haskell values]
+The first thought that comes to mind when you asked to convert a PLC value to the corresponding Haskell
+value is "just match on the AST". This works nicely for simple things like 'Char's which we encode as
+@integer@s, see the @KnownDynamicBuiltinType Char@ instance below.
+
+But how to convert something more complicated like lists? A PLC list gets passed as argument to
+a built-in after it gets evaluated to WHNF. We can't just match on the AST here, because after
+the initial lambda it can be anything there: function applications, other built-ins, recursive data,
+anything. "Well, just normalize it" -- not so fast: for one, we did not have a term normalization
+procedure at the moment this note was written, for two, it's not something that can be easily done,
+because you have to carefully handle uniques (we generate new terms during evaluation) and perform type
+substitutions, because types must be preserved.
+
+Besides, matching on the AST becomes really complicated: you have to ensure that a term does have
+an expected semantics by looking at the term's syntax. Huge pattern matches followed by multiple
+checks that variables have equal names in right places and have distinct names otherwise. Making a
+mistake is absolutely trivial here. Of course, one could just omit checks and hope it'll work alright,
+but eventually it'll break and debugging won't be fun at all.
+
+So instead of dealing with syntax of terms, we deal with their semantics. Namely, we evaluate terms
+using some evaluator (normally, the CEK machine). For the temporary lack of ability to put values of
+arbitrary Haskell types into the Plutus Core AST, we convert PLC values to Haskell values and "emit"
+the latter via a combination of 'unsafePerformIO' and 'IORef'. For example, we fold a PLC list with
+a dynamic built-in name (called `emit`) that calls 'unsafePerformIO' over a Haskell function that
+appends an element to the list stored in an 'IORef':
+
+    plcListToHaskellList list =
+        evaluateCek anEnvironment (foldList {dyn} {unit} (\(r : unit) -> emit) unitval list)
+
+After evaluation finishes, we read a Haskell list from the 'IORef'
+(which requires another 'unsafePerformIO') and return it.
```

Comment:
What would happen if we made the monad that the CEK machine runs in a monad transformer, and allowed builtin evaluation to have effects in that monad?

Then, for example, `emit` could use a `STRef` instead of an `IORef` (or indeed we could keep using an `IORef` and run over IO, but that seems overkill). This would also be a way of allowing other kinds of effects inside builtins.

This is a very sketchy suggestion so far, it's possible that the types make this difficult. I think it would be okay if we instantiated our monad differently in different cases: so for the "main" CEK evaluation it's `Identity`, but for these `evaluateCek` calls you use to convert the builtins to Haskell values you could use `ST`.

Reply:
Interesting thoughts. I'll probably play with this stuff later. But we need some motivation for it. `emit` is a hack that is supposed to vanish at some point. Do we need effects for something else?

Reply:
Well, I'm pretty uncomfortable with the `unsafePerformIO` stuff in here, so I'd feel a lot more like I understood what was going on and that I wasn't going to accidentally break things by subtly changing how GHC complies things.

In addition, I think we're planning to keep the trace effect in the mockchain for a while, so that could be modelled as happening in an appropriate base monad

Reply:
> Well, I'm pretty uncomfortable with the `unsafePerformIO` stuff in here

So am I, but I have only two days left and the main stuff is not done yet. I'll happily try out your suggestion later.

Reply:
Well, okay. Maybe I'll try it while you're away.

Reply:
That's not fair!

Reply:
I can't let you have all the fun :p